```
error[E0424]: expected value, found module `self`

--> /home/travis/build/servo/servo/components/hashglobe/src/hash_map.rs:1344:9

|

1344 | self.inner.next_with_hash()

| ^^^^ `self` value is only available in methods with `self` parameter

error[E0599]: no method named `next_with_hash` found for type `hash_map::Iter<'_, K, (usize, V)>` in the current scope

--> /home/travis/build/servo/servo/components/hashglobe/src/diagnostic.rs:67:42

|

67 | while let Some((h, _, v)) = iter.next_with_hash() {

| ^^^^^^^^^^^^^^

|

= note: found the following associated functions; to be used as methods, functions must have a `self` parameter

= help: try with `hash_map::Iter<'_, K, (usize, V)>::next_with_hash`

note: candidate #1 is defined in an impl for the type `hash_map::Iter<'_, _, _>`

--> /home/travis/build/servo/servo/components/hashglobe/src/hash_map.rs:1343:5

|

134| / fn next_with_hash() -> Option<(usize, &'a K, &'a V)> {

134| | self.inner.next_with_hash()

134| | }

| |_____^

error: aborting due to 2 previous errors
```

--- a/components/hashglobe/src/diagnostic.rs
+++ b/components/hashglobe/src/diagnostic.rs
@@ -62,17 +62,21 @@ impl<K: Hash + Eq, V, S: BuildHasher> DiagnosticHashMap<K, V, S>
         let mut position = 0;
         let mut count = 0;
         let mut bad_canary = None;
-        for (_,v) in self.map.iter() {
+
+        let mut iter = self.map.iter();
+        while let Some((h, _, v)) = iter.next_with_hash() {
             let canary_ref = &v.0;
             position += 1;
+
             if *canary_ref == CANARY {
                 continue;
             }
+
             count += 1;
-            bad_canary = Some((*canary_ref, canary_ref, position));
+            bad_canary = Some((h, *canary_ref, canary_ref, position));
         }
         if let Some(c) = bad_canary {
-            self.report_corruption(c.0, c.1, c.2, count);
+            self.report_corruption(c.0, c.1, c.2, c.3, count);
         }
     }
 
@@ -158,6 +162,7 @@ impl<K: Hash + Eq, V, S: BuildHasher> DiagnosticHashMap<K, V, S>
     #[inline(never)]
     fn report_corruption(
         &self,
+        hash: usize,
         canary: usize,
         canary_addr: *const usize,
         position: usize,
@@ -172,12 +177,14 @@ impl<K: Hash + Eq, V, S: BuildHasher> DiagnosticHashMap<K, V, S>
                 value.as_ptr(),
             );
         }
+
         panic!(
-            concat!("HashMap Corruption (sz={}, cap={}, pairsz={}, cnry={:#x}, count={}, ",
-                    "last_pos={}, base_addr={:?}, cnry_addr={:?}, jrnl_len={})"),
+            concat!("HashMap Corruption (sz={}, cap={}, pairsz={}, hash={:#x}, cnry={:#x}, ",
+                "count={}, last_pos={}, base_addr={:?}, cnry_addr={:?}, jrnl_len={})"),
             self.map.len(),
             self.map.raw_capacity(),
             ::std::mem::size_of::<(K, (usize, V))>(),
+            hash,
             canary,
             count,
             position,

--- a/components/hashglobe/src/hash_map.rs
+++ b/components/hashglobe/src/hash_map.rs
@@ -1339,6 +1339,12 @@ impl<'a, K: Debug, V: Debug> fmt::Debug for Iter<'a, K, V> {
     }
 }
 
+impl<'a, K: 'a, V: 'a>  Iter<'a, K, V> {
+    pub fn next_with_hash(&mut self) -> Option<(usize, &'a K, &'a V)> {
+        self.inner.next_with_hash()
+    }
+}
+
 /// A mutable iterator over the entries of a `HashMap`.
 ///
 /// This `struct` is created by the [`iter_mut`] method on [`HashMap`]. See its

--- a/components/hashglobe/src/table.rs
+++ b/components/hashglobe/src/table.rs
@@ -1130,6 +1130,15 @@ impl<'a, K, V> ExactSizeIterator for Iter<'a, K, V> {
     }
 }
 
+impl<'a, K, V> Iter<'a, K, V> {
+    pub fn next_with_hash(&mut self) -> Option<(usize, &'a K, &'a V)> {
+        self.iter.next().map(|raw| unsafe {
+            let (hash_ptr, pair_ptr) = raw.hash_pair();
+            (*hash_ptr, &(*pair_ptr).0, &(*pair_ptr).1)
+        })
+    }
+}
+
 impl<'a, K, V> Iterator for IterMut<'a, K, V> {
     type Item = (&'a K, &'a mut V);
 

